tests/test_training/test_opt.py

```diff
@@ -192,65 +187,100 @@ def cost_fn():
 def test_learning_four_mode_Interferometer():
     """Finding the optimal Interferometer to make a NOON state with N=2"""
     settings.SEED = 4
+    solution_U = np.array(
+        [
+            [
+                -0.47541806 + 0.00045878j,
+                -0.41513474 - 0.27218387j,
+                -0.11065812 - 0.39556922j,
+                -0.29912017 + 0.51900235j,
+            ],
+            [
+                -0.05246398 + 0.5209089j,
+                -0.29650069 - 0.40653082j,
+                0.57434638 - 0.04417284j,
+                0.28230532 - 0.24738672j,
+            ],
+            [
+                0.28437557 + 0.08773767j,
+                0.18377764 - 0.66496587j,
+                -0.5874942 - 0.19866946j,
+                0.2010813 - 0.10210844j,
+            ],
+            [
+                -0.63173183 - 0.11057324j,
+                -0.03468292 + 0.15245454j,
+                -0.25390362 - 0.2244298j,
+                0.18706333 - 0.64375049j,
+            ],
+        ]
+    )
+    perturbed = (
+        Interferometer(num_modes=4, unitary=solution_U)
+        >> BSgate(settings.rng.normal(scale=0.01), modes=[0, 1])
+        >> BSgate(settings.rng.normal(scale=0.01), modes=[2, 3])
+        >> BSgate(settings.rng.normal(scale=0.01), modes=[1, 2])
+        >> BSgate(settings.rng.normal(scale=0.01), modes=[0, 3])
+    )
+    perturbed_U = perturbed._ops[0].unitary.value
+
     ops = [
         Sgate(
-            r=settings.rng.uniform(size=4),
-            phi=settings.rng.normal(size=4),
+            r=settings.rng.normal(loc=np.arcsinh(1.0), scale=0.01, size=4),
             r_trainable=True,
-            phi_trainable=True,
         ),
-        Interferometer(num_modes=4, unitary_trainable=True),
+        Interferometer(unitary=perturbed_U, num_modes=4, unitary_trainable=True),
     ]
     circ = Circuit(ops)
-    state_in = Vacuum(num_modes=4)
 
     def cost_fn():
-        amps = (state_in >> circ).ket(cutoffs=[3, 3, 3, 3])
-        return (
-            -math.abs(
-                math.sum(
-                    amps[1, 1]
-                    * np.array([[0, 0, 1 / np.sqrt(2)], [0, 0, 0], [1 / np.sqrt(2), 0, 0]])
-                )
-            )
-            ** 2
-        )
+        amps = (Vacuum(num_modes=4) >> circ).ket(cutoffs=[3, 3, 3, 3])
+        return -math.abs((amps[1, 1, 2, 0] + amps[1, 1, 0, 2]) / np.sqrt(2)) ** 2
 
     opt = Optimizer(unitary_lr=0.05)
-    opt.minimize(cost_fn, by_optimizing=[circ], max_steps=1000)
+    opt.minimize(cost_fn, by_optimizing=[circ], max_steps=200)
     assert np.allclose(-cost_fn(), 0.0625, atol=1e-5)
 
 
 def test_learning_four_mode_RealInterferometer():
     """Finding the optimal Interferometer to make a NOON state with N=2"""
     settings.SEED = 6
+    solution_O = np.array(
```

Comment:
Is this the numerical solution or an analytical solution?

Reply:
analytical